checker/src/main/java/org/checkerframework/checker/mustcall/MustCallVisitor.java

```diff
@@ -0,0 +1,193 @@
+import com.sun.source.tree.MethodInvocationTree;
+import com.sun.source.tree.MethodTree;
+import com.sun.source.tree.ReturnTree;
+import com.sun.source.tree.Tree;
+import java.util.Arrays;
+import java.util.Collections;
+import java.util.Set;
+import javax.lang.model.element.AnnotationMirror;
+import javax.lang.model.element.ElementKind;
+import javax.lang.model.element.ExecutableElement;
+import org.checkerframework.checker.compilermsgs.qual.CompilerMessageKey;
+import org.checkerframework.checker.mustcall.qual.NotOwning;
+import org.checkerframework.common.basetype.BaseTypeChecker;
+import org.checkerframework.common.basetype.BaseTypeVisitor;
+import org.checkerframework.common.basetype.TypeValidator;
+import org.checkerframework.framework.type.AnnotatedTypeMirror;
+import org.checkerframework.framework.type.AnnotatedTypeMirror.AnnotatedExecutableType;
+import org.checkerframework.javacutil.ElementUtils;
+import org.checkerframework.javacutil.TreePathUtil;
+import org.checkerframework.javacutil.TreeUtils;
+
+/**
+ * The visitor for the Must Call checker. This visitor is similar to BaseTypeVisitor, but overrides
+ * methods that don't work well with the MustCall type hierarchy because it doesn't use the top type
+ * as the default type.
+ */
+public class MustCallVisitor extends BaseTypeVisitor<MustCallAnnotatedTypeFactory> {
+
+  /**
+   * Creates a new MustCallVisitor.
+   *
+   * @param checker the type-checker associated with this visitor
+   */
+  public MustCallVisitor(BaseTypeChecker checker) {
+    super(checker);
+  }
+
+  @Override
+  public Void visitReturn(ReturnTree node, Void p) {
+    // Only check return types if ownership is being transferred.
+    MethodTree enclosingMethod = TreePathUtil.enclosingMethod(this.getCurrentPath());
+    // enclosingMethod is null if this return site is inside a lambda. TODO: handle lambdas more
+    // precisely?
+    if (!checker.hasOption(MustCallChecker.NO_LIGHTWEIGHT_OWNERSHIP) && enclosingMethod != null) {
+      ExecutableElement methodElt = TreeUtils.elementFromDeclaration(enclosingMethod);
+      AnnotationMirror notOwningAnno = atypeFactory.getDeclAnnotation(methodElt, NotOwning.class);
+      if (notOwningAnno != null) {
+        // skip return type subtyping check, because not-owning pointer means OCC won't check anyway
+        return null;
+      }
+    }
+    return super.visitReturn(node, p);
+  }
+
+  @Override
+  protected boolean skipReceiverSubtypeCheck(
+      MethodInvocationTree node,
+      AnnotatedTypeMirror methodDefinitionReceiver,
+      AnnotatedTypeMirror methodCallReceiver) {
+    // TODO: Check explicit receiver parameters annotated with @Owning. ExecutableElement
+    //       doesn't have any way to get an element associated with the receiver, so I can't
+    //       figure out a way to get a declaration annotation for the receiver. It might not
+    //       be possible? The below is the closest that I got, but the receiver doesn't show up
```

Comment:
It turns out that `@Owning` receiver parameters don't make sense anyway, as I mentioned in one of my other responses to your comments. So I'm just going to delete this entire comment.

Thanks for these references, though - this will be useful the next time I try to figure out why I can't write a declaration annotation on a receiver and then access it from a checker :)